File: aicode/minicli4/typescript/src/session.ts
import fs from "node:fs";
import path from "node:path";

export type MessageRole = "system" | "user" | "assistant" | "tool";

export interface SessionMessage {
  role: MessageRole;
  content: string;
  created_at: string;
  tool_call_id?: string;
  name?: string;
}

export interface SessionRecord {
  session_id: string;
  title: string;
  messages: SessionMessage[];
  updated_at: string;
}

function sessionsDir(stateDir: string): string {
  return path.join(stateDir, "sessions");
}

function currentPath(stateDir: string): string {
  return path.join(stateDir, "current.txt");
}

function sessionPath(stateDir: string, id: string): string {
  return path.join(sessionsDir(stateDir), `${id}.json`);
}

function nowIso(): string {
  return new Date().toISOString();
}

export class SessionStore {
  constructor(private readonly stateDir: string) {}

  private ensure(): void {
    fs.mkdirSync(sessionsDir(this.stateDir), { recursive: true });
  }

  list(): SessionRecord[] {
    this.ensure();
    const files = fs.readdirSync(sessionsDir(this.stateDir)).filter((x) => x.endsWith(".json"));
    return files.map((name) => {
      const full = path.join(sessionsDir(this.stateDir), name);
      const rec = JSON.parse(fs.readFileSync(full, "utf8")) as SessionRecord;
      return rec;
    }).sort((a, b) => b.updated_at.localeCompare(a.updated_at));
  }

  create(id?: string): SessionRecord {
    this.ensure();
    const sessionId = id && id.trim().length > 0 ? id.trim() : `session-${Date.now()}`;
    const rec: SessionRecord = {
      session_id: sessionId,
      title: sessionId,
      messages: [],
      updated_at: nowIso()
    };
    this.save(rec);
    this.setCurrent(sessionId);
    return rec;
  }

  load(id: string): SessionRecord {
    const full = sessionPath(this.stateDir, id);
    if (!fs.existsSync(full)) {
      throw new Error(`session not found: ${id}`);
    }
    return JSON.parse(fs.readFileSync(full, "utf8")) as SessionRecord;
  }

  save(rec: SessionRecord): void {
    this.ensure();
    rec.updated_at = nowIso();
    fs.writeFileSync(sessionPath(this.stateDir, rec.session_id), JSON.stringify(rec, null, 2), "utf8");
  }

  setCurrent(id: string): void {
    this.ensure();
    fs.writeFileSync(currentPath(this.stateDir), id, "utf8");
  }

  current(): SessionRecord {
    this.ensure();
    const marker = currentPath(this.stateDir);
    if (!fs.existsSync(marker)) {
      return this.create("default");
    }
    const id = fs.readFileSync(marker, "utf8").trim();
    try {
      return this.load(id);
    } catch {
      return this.create("default");
    }
  }

  delete(id: string): boolean {
    const full = sessionPath(this.stateDir, id);
    if (!fs.existsSync(full)) {
      return false;
    }
    fs.unlinkSync(full);
    return true;
  }

  rename(oldId: string, newId: string): boolean {
    const oldPath = sessionPath(this.stateDir, oldId);
    const newPath = sessionPath(this.stateDir, newId);
    if (!fs.existsSync(oldPath) || fs.existsSync(newPath)) {
      return false;
    }
    const rec = this.load(oldId);
    rec.session_id = newId;
    rec.title = newId;
    fs.unlinkSync(oldPath);
    this.save(rec);
    if (this.current().session_id === oldId) {
      this.setCurrent(newId);
    }
    return true;
  }
}

File: aicode/minicli4/cpp/src/session.cpp
#include "minicli4/session.hpp"

#include <algorithm>
#include <chrono>
#include <iomanip>
#include <sstream>
#include <stdexcept>

namespace minicli4 {

std::string now_iso() {
  const auto now = std::chrono::system_clock::now();
  const std::time_t t = std::chrono::system_clock::to_time_t(now);
  std::tm tm{};
#if defined(_WIN32)
  gmtime_s(&tm, &t);
#else
  gmtime_r(&t, &tm);
#endif
  std::ostringstream oss;
  oss << std::put_time(&tm, "%Y-%m-%dT%H:%M:%SZ");
  return oss.str();
}

SessionRecord SessionStore::create(const std::string& session_id) {
  std::string sid = session_id;
  if (sid.empty()) {
    sid = "session-" + std::to_string(static_cast<long long>(
                           std::chrono::duration_cast<std::chrono::milliseconds>(
                               std::chrono::system_clock::now().time_since_epoch())
                               .count()));
  }
  SessionRecord rec{sid, sid, {}, now_iso()};
  records_[sid] = rec;
  current_id_ = sid;
  return rec;
}

SessionRecord SessionStore::load(const std::string& session_id) const {
  const auto it = records_.find(session_id);
  if (it == records_.end()) {
    throw std::runtime_error("session not found: " + session_id);
  }
  return it->second;
}

void SessionStore::save(const SessionRecord& rec) {
  SessionRecord next = rec;
  next.updated_at = now_iso();
  records_[next.session_id] = next;
}

std::vector<SessionRecord> SessionStore::list() const {
  std::vector<SessionRecord> out;
  out.reserve(records_.size());
  for (const auto& [_, rec] : records_) {
    out.push_back(rec);
  }
  std::sort(out.begin(), out.end(), [](const SessionRecord& a, const SessionRecord& b) {
    return a.updated_at > b.updated_at;
  });
  return out;
}

void SessionStore::set_current(const std::string& session_id) {
  current_id_ = session_id;
}

SessionRecord SessionStore::current() const {
  if (current_id_.empty()) {
    auto* self = const_cast<SessionStore*>(this);
    return self->create("default");
  }
  const auto it = records_.find(current_id_);
  if (it == records_.end()) {
    auto* self = const_cast<SessionStore*>(this);
    return self->create("default");
  }
  return it->second;
}

bool SessionStore::remove(const std::string& session_id) {
  return records_.erase(session_id) > 0;
}

bool SessionStore::rename(const std::string& old_id, const std::string& new_id) {
  if (records_.count(old_id) == 0 || records_.count(new_id) > 0) {
    return false;
  }
  SessionRecord rec = records_.at(old_id);
  rec.session_id = new_id;
  rec.title = new_id;
  rec.updated_at = now_iso();
  records_.erase(old_id);
  records_[new_id] = rec;
  if (current_id_ == old_id) {
    current_id_ = new_id;
  }
  return true;
}

}  // namespace minicli4
